src/index.ts

```diff
@@ -12,42 +12,91 @@ import { internalPubkey } from "./constants/internalPubkey";
 import { initBTCCurve } from "./utils/curve";
 import { PK_LENGTH, StakingScriptData } from "./utils/stakingScript";
 import { UTXO } from "./types/UTXO";
+import { getEstimatedFee, inputValueSum } from "./utils/fee";
 
 export { initBTCCurve, StakingScriptData };
 
 // https://bips.xyz/370
 const BTC_LOCKTIME_HEIGHT_TIME_CUTOFF = 500000000;
+const BTC_DUST_SAT = 546;
+
+export interface PsbtTransactionResult {
+  psbt: Psbt;
+  fee: number;
+}
+
+const getStakingTxInputUTXOsAndFees = (
+  availableUTXOs: UTXO[],
+  stakingAmount: number,
+  feeRate: number,
+  numOfOutputs: number,
+): {
+  selectedUTXOs: UTXO[],
+  fee: number,
+} => {
+  if (availableUTXOs.length === 0) {
+    throw new Error("Insufficient funds");
+  }
+  // Sort available UTXOs from highest to lowest value
+  availableUTXOs.sort((a, b) => b.value - a.value);
+
+  let selectedUTXOs: UTXO[] = [];
+  let accumulatedValue = 0;
+  let estimatedFee = getEstimatedFee(feeRate, 1, numOfOutputs); // Initial fee estimate with 1 input
+
+  for (const utxo of availableUTXOs) {
+    selectedUTXOs.push(utxo);
+    accumulatedValue += utxo.value;
+    estimatedFee = getEstimatedFee(feeRate, selectedUTXOs.length, numOfOutputs);
+
+    if (accumulatedValue >= stakingAmount + estimatedFee) {
+      break;
+    }
+  }
```

Comment:
```suggestion
  // Can be moved to the top of the function
  if (len(availableUTXOs) === 0) {
    throw new Error("No input UTXOs provided");
  }
  
  // Can probs use a more native JS way
  let idx = 0;
  let selectedUTXOs: UTXO[] = [availableUTXOs[idx]];
  let accumulatedValue = availableUTXOs[0].value;
  
  while (accumulatedValue < stakingAmount + getEstimatedFee(feeRate, selectedUTXOs.length, numOfOutputs) {
    idx += 1;
    if (idx >= availableUTXOs.length) {
      throw new Error("Insufficient funds: unable to gather enough UTXOs to cover the staking amount and fees.");
    }
    accumulatedValue += availableUTXOs[idx].value;
    selectedUTXOs.push(availableUTXOs[idx]);
  }
  return {
    selectedUTXOs,
    fee: estimatedFee,
  };
```

It is a bit weird that we initialize `selectedUTXOs` to an empty list, but the initial estimated fee is with a single input.

How about we start with the `selectedUTXOs` list containing the first UTXO and then we have a `while` loop that adds one more UTXO until `accumulatedInputsValue >= stakingAmount + getEstimatedFee(feeRate, selectedUTXOs.length, numOfOutputs)`? Like the semi pseudo-code above.

Reply:
i generally don't suggest to use while when there is alternatives. in this case that we already have `availableUTXOs` so ideal methods would be iterating through the availableUTXOs directly using `.map, .reduce, for, forEach` etc. `while` is a least preferred option as it subject to infinite loops if not doing it right.

In the current implementation, that you pointed out. I don't think it's the issue of `selectedUTXOs` being empty. But instead, we can simply remove the `let estimatedFee = getEstimatedFee(feeRate, 1, numOfOutputs); // Initial fee estimate with 1 input` all together. there is no use of the initial estimated fee in the methods anymore. the for loop will take care of it for each inputs we add into the selectedInputs